carma_cooperative_perception/src/msg_conversion.cpp

```diff
@@ -265,4 +363,148 @@ auto to_external_object_list_msg(
+
+    detected_object_common_data.presence_vector |= carma_v2x_msgs::msg::DetectedObjectCommonData::HAS_SPEED_Z;
+    detected_object_common_data.speed_z.speed = external_object.velocity_inst.twist.linear.z;
+
+    // heading - convert ang vel to scale heading
+    detected_object_common_data.heading.heading = external_object.velocity_inst.twist.angular.z * (180.0/3.14115926535); // to radian, use units.h
+  }
+
+  // optional data (determine based on object type)
+  // use object type struct for better control
+  carma_v2x_msgs::msg::DetectedObjectOptionalData detected_object_optional_data;
+
+  switch (external_object.object_type) {
+    case external_object.SMALL_VEHICLE:
+      detected_object_common_data.obj_type.object_type |= j3224_v2x_msgs::msg::ObjectType::VEHICLE;
+
+      if(external_object.presence_vector & external_object.SIZE_PRESENCE_VECTOR){
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_SIZE;
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_HEIGHT;
+
+        detected_object_optional_data.det_veh.size.vehicle_width = external_object.size.y;
+        detected_object_optional_data.det_veh.size.vehicle_length = external_object.size.x;
+        detected_object_optional_data.det_veh.height.vehicle_height = external_object.size.z;
+      }
+      break;
+    case external_object.LARGE_VEHICLE:
+      detected_object_common_data.obj_type.object_type |= j3224_v2x_msgs::msg::ObjectType::VEHICLE;
+
+      if(external_object.presence_vector & external_object.SIZE_PRESENCE_VECTOR){
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_SIZE;
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_HEIGHT;
+
+        detected_object_optional_data.det_veh.size.vehicle_width = external_object.size.y;
+        detected_object_optional_data.det_veh.size.vehicle_length = external_object.size.x;
+        detected_object_optional_data.det_veh.height.vehicle_height = external_object.size.z;
+      }
+      break;
+    case external_object.MOTORCYCLE:
+      detected_object_common_data.obj_type.object_type |= j3224_v2x_msgs::msg::ObjectType::VEHICLE;
+
+      if(external_object.presence_vector & external_object.SIZE_PRESENCE_VECTOR){
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_SIZE;
+        detected_object_optional_data.det_veh.presence_vector |= carma_v2x_msgs::msg::DetectedVehicleData::HAS_HEIGHT;
+
+        detected_object_optional_data.det_veh.size.vehicle_width = external_object.size.y;
+        detected_object_optional_data.det_veh.size.vehicle_length = external_object.size.x;
+        detected_object_optional_data.det_veh.height.vehicle_height = external_object.size.z;
```

Comment:
These all look identical. Can the common steps be moved to a function to avoid duplicate code?

Reply:
I also don't love this but the alternative seems to be a messy if statement which doesn't mesh as well with the enumeration-like implementation of object_type. Preference, suggestion?